### PR TITLE
Type `this` in more constructor functions

### LINKED ISSUE #18171: In JS, constructor functions without property assignments aren't recognised as classes
**TypeScript Version:**  2.5.1

**Code**

For the JS:

```js
function Foo() {
	this.init();
}

Foo.prototype.init = function () {
	// code
}

const a = new Foo();
a.init()
```

**Expected behavior:**
In the function `Foo`, `this` is the type of `Foo` and you can run `go to definition` on `this.init`

**Actual behavior:**
`this` is `any`. `go to definition` does not work on `this.init` but does on `a.init`

### PR DESCRIPTION
**Pull request by sandersn**
Previously,  `this: this` in constructor functions only when there was an explicit `@constructor` tag on the function. Now, `this: this` for any function that's known to be a constructor function.

This improves completions inside constructor functions; also note that previously the compiler *did* type `this: this` inside methods of constructor functions, so this fix makes us more consistent. This is reflected in the large number of baselines that improve.

The fix is a simple switch to `isJSConstructor`, which is the standard way to detect constructor functions. I'm not sure why the original PR didn't use this method.

I remember discussing this limitation in the original bug, #25979, and I guess I decided that it made sense. But I was heavily primed by the bug's framing of the problem in terms of `noImplicitThis`, which *should* require an explicit `@constructor` tag.

- With better typing comes better detection of `@readonly` assignment; I had to fix the readonly detection code to use `isJSConstructor` as well.
- Also, previously, both `@class` and `@this` in a jsdoc would cause the `@this` annotation to be ignored. This became a worse problem with this PR, because `this` is correctly typed even without the annotation. This PR now makes sure that `@this` is checked first and used if present.

Fixes #18171

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -22416,30 +22416,21 @@ namespace ts {
...
-                // Check if it's a constructor definition, can be either a variable decl or function decl
-                // i.e.
-                //   * /** @constructor */ function [name]() { ... }
-                //   * /** @constructor */ var x = function() { ... }
-                else if (isInJS &&
-                         (container.kind === SyntaxKind.FunctionExpression || container.kind === SyntaxKind.FunctionDeclaration) &&
-                         getJSDocClassTag(container)) {
-                    const classType = (getDeclaredTypeOfSymbol(getMergedSymbol(container.symbol)) as InterfaceType).thisType!;
-                    return getFlowTypeOfReference(node, classType);
+                else if (isJSConstructor(container)) {
```
Comment (sandersn) [inline src/compiler/checker.ts:22429]: this is the real change of the PR; the rest just dedupes `getFlowTypeOfReference` calls to the bottom of the function.

**Comment by sandersn** [comment]
woops, missed some fourslash tests since those don't get baseline updated.

**Comment by sandersn** [comment]
The fourslash tests point out a codefix that doesn't apply anymore; with this change `this` is usually typed in constructor functions, so explicit marking with `@class` is less often needed.

I'm not sure if this contradicts the original spirit of #25979, in which we decided that `noImplicitThis: true` implies that the codebase wants to make constructor functions less convenient by forcing annotation, but it's way more convenient, and it's useful in the usual case where `noImplicitThis: false` in JS.

It's possible to get back the original `noImplicitThis: true` error+codefix by issuing a noImplicitThis-only error while still returning the correct type, but I don't really think it's worth the effort.

**Comment by sandersn** [comment]
Now master is broken with some lint. I'll merge it in again once it's fixed.

**Comment by sandersn** [comment]
@rachelgshaffer you might be interested in this as well, especially now that it changes a codefix.

**Comment by sandersn** [comment]
In fact, it looks like I imagined requesting reviewers at all, because the reviewer list is broken. @andrewbranch @elibarzilay can you take a look?

**Comment by sandersn** [comment]
Since it only affects JS files, for non-checkJS compiles, nobody will notice anything. For checkJS compiles, which are a lot rarer, I think the improvement is worth it.

Also, the `this` type is used everywhere except constructor function bodies, so it's not like we're exposing a new code path to the world.

However, it's a good idea to check this on the user tests...

@typescript-bot user test this

**Comment by andrewbranch** [comment]
> Also, the `this` type is used everywhere except constructor function bodies, so it's not like we're exposing a new code path to the world.

Ah, I didn’t realize that. That makes it seem like a smaller change then.

**Comment by sandersn** [comment]
Briefly: yes, it is pretty breaky. I'm going to see what the breaks look like before merging.

**Comment by sandersn** [comment]
acorn: acorn's distributed code uses a weird pattern that aliases the prototype like so

```js
let pp$1 = Parser.prototype
pp$1.method = function() { ... }
```

Then any method with a this-property assignment, which is most of them, treats the method as a fresh constructor function instead.

It's fine that we don't understand that pattern, but this problem will happen anywhere that we misunderstand a method as a constructor function.

**Comment by sandersn** [comment]
The `@this` bug is actually an existing conflict that you can expose today with

```js
/**
 * @class
 * @this {{ e: 1, m: 1 }}
 */
function C() {
  this.e = this.m + 1
}
```

I think it's pretty easy to fix, though.

**Comment by sandersn** [comment]
@typescript-bot user test this

**Comment by sandersn** [comment]
The `@this` fix improves devtools-frontend's errors, but adds some more errors where the code extends built-ins with an explicit `@this` comment that we previously ignored. I'm going to look at the rest of the new errors now.

**Comment by sandersn** [comment]
So, to summarise: There are lots of new errors, mostly in patterns we already didn't understand, or attempt to understand. Before, `this: any`, but now `this: this`.

I think this is fine because the typical case of `checkJs: false` will just get improved completions, and people with `checkJs: true` are likely to use newer constructs like classes, based on survey data from 2018 at least.

**Pull request merged**